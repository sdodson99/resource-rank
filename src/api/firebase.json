{
  "functions": {
    "source": "."
  },
  "emulators": {
    "functions": {
      "port": 5001
    },
    "database": {
      "port": 9000
    },
    "auth": {
      "port": 9099
    },
    "ui": {
      "enabled": true
    }
  }
}
