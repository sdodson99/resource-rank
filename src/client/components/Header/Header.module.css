.Header {
  background-color: var(--color-primary);

  @apply py-5 text-white drop-shadow-md;
}

.NavItem {
  @apply sm:pl-8 pt-3 sm:pt-0;
}

.NavLink {
  @apply text-xl text-white no-underline opacity-80 hover:opacity-100 transition duration-200;
}

.ActiveNavLink {
  @apply opacity-100;
}
