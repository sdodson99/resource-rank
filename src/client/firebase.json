{
  "hosting": {
    "public": "public",
    "ignore": ["firebase.json", "**/.*", "**/node_modules/**"],
    "rewrites": [
      {
        "source": "**",
        "function": "nextjs"
      }
    ]
  },
  "functions": {
    "source": ".",
    "runtime": "nodejs12"
  },
  "emulators": {
    "hosting": {
      "port": 5000
    },
    "functions": {
      "port": 7001
    },
    "ui": {
      "enabled": true
    }
  }
}
