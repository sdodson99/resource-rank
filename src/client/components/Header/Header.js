import React, { useState } from 'react';
import styles from './Header.module.css';
import Link from 'next/link';
import LoadingButton from '../LoadingButton/LoadingButton';
import Image from 'next/image';
import firebase from 'firebase/app';
import useAuthenticationContext from '../../hooks/authentication/use-authentication-context';
import useFirebaseApp from '../../hooks/use-firebase-app';
import { useRouter } from 'next/router';

const Header = () => {
  const router = useRouter();
  const firebaseApp = useFirebaseApp();
  const { isLoggedIn } = useAuthenticationContext();
  const [isLoggingIn, setIsLoggingIn] = useState(false);
  const [isLoggingOut, setIsLoggingOut] = useState(false);

  const onLoginClick = async () => {
    setIsLoggingIn(true);

    try {
      const googleAuthProvider = new firebase.auth.GoogleAuthProvider();

      await firebaseApp.auth().signInWithPopup(googleAuthProvider);
    } catch (error) {
      console.log(error);
    } finally {
      setIsLoggingIn(false);
    }
  };

  const onLogoutClick = async () => {
    setIsLoggingOut(true);

    try {
      await firebaseApp.auth().signOut();
    } catch (error) {
      console.log(error);
    } finally {
      setIsLoggingOut(false);
    }
  };

  const getLinkClassName = (href) => {
    let className = `${styles.NavItem} ${styles.NavLink}`;

    if (href === router.route) {
      className += ` ${styles.Active}`;
    }

    return className;
  };

  return (
    <header className={styles.Header} data-testid="Header">
      <div className="content-container">
        <div className="sm:flex items-center text-center sm:text-left">
          <div className="flex-grow">
            <Link href="/">
              <a>
                <Image
                  src="/img/logo.svg"
                  alt="Resource Rank Logo"
                  height="50"
                  width="50"
                  layout="fixed"
                  priority={true}
                />
              </a>
            </Link>
          </div>

          <div className={getLinkClassName('/')}>
            <Link href="/">
              <a>Home</a>
            </Link>
          </div>

          <div className={getLinkClassName('/topics')}>
            <Link href="/topics">
              <a>Topics</a>
            </Link>
          </div>

          <div className={styles.NavItem}>
            {!isLoggedIn && (
              <LoadingButton isLoading={isLoggingIn} onClick={onLoginClick}>
                Login
              </LoadingButton>
            )}
            {isLoggedIn && (
              <LoadingButton isLoading={isLoggingOut} onClick={onLogoutClick}>
                Logout
              </LoadingButton>
            )}
          </div>
        </div>
      </div>
    </header>
  );
};

Header.propTypes = {};

Header.defaultProps = {};

export default Header;
