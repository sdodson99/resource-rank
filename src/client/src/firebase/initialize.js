import firebase from 'firebase/app';
import 'firebase/auth';

let firebaseInstance;

export default function initialize(
  firebaseConfig,
  { useAuthenticationEmulator }
) {
  if (firebaseInstance) {
    return firebaseInstance;
  }

  if (typeof window === 'undefined') {
    return null;
  }

  firebaseInstance = firebase.initializeApp(firebaseConfig);

  const auth = firebase.auth();
  if (useAuthenticationEmulator) {
    auth.useEmulator('http://localhost:9099');
  }

  return firebaseInstance;
}
