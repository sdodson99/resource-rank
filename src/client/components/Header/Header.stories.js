import React from 'react';
import Header from './Header';

export default {
  component: Header,
  title: 'Components/Header',
  decorators: [
    (Story) => {
      return <Story />;
    },
  ],
};

const Template = (args) => <Header {...args} />;

export const Default = Template.bind({});
Default.args = {};
