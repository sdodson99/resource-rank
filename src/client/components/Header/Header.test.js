import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import Header from './Header';
import { useRouter } from 'next/router';
import useFirebaseAppContext from '../../hooks/use-firebase-app-context';
import useFirebaseAnalytics from '../../hooks/use-firebase-analytics';
import useAuthenticationContext from '../../hooks/use-authentication-context';
import 'firebase/auth';

jest.mock('../../hooks/use-authentication-context');
jest.mock('../../hooks/use-firebase-app-context');
jest.mock('../../hooks/use-firebase-analytics');
jest.mock('next/router');

describe('<Header />', () => {
  let mockFirebaseApp;
  let mockSignIn;
  let mockSignOut;
  let mockConsoleError;
  let originalConsoleError;

  beforeEach(() => {
    useRouter.mockReturnValue({
      route: '/',
    });

    mockSignIn = jest.fn();
    mockSignOut = jest.fn();
    mockFirebaseApp = {
      auth: () => ({
        signInWithPopup: mockSignIn,
        signOut: mockSignOut,
      }),
    };
    useFirebaseAppContext.mockReturnValue(mockFirebaseApp);
    useFirebaseAnalytics.mockReturnValue({
      logEvent: jest.fn(),
    });

    useAuthenticationContext.mockReturnValue({ isLoggedIn: false });

    mockConsoleError = jest.fn();
    originalConsoleError = console.error;
    console.error = mockConsoleError;
  });

  afterEach(() => {
    useRouter.mockReset();
    useFirebaseAppContext.mockReset();
    useFirebaseAnalytics.mockReset();
    useAuthenticationContext.mockReset();
    console.error = originalConsoleError;
  });

  it('should mount', () => {
    render(<Header />);

    const header = screen.getByTestId('Header');

    expect(header).toBeInTheDocument();
  });

  describe('with not logged in', () => {
    it('should render correctly', () => {
      const tree = renderer.create(<Header />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should login when login button clicked', () => {
      render(<Header />);
      const loginButton = screen.getByText('Login');

      loginButton.click();

      expect(mockSignIn).toBeCalled();
    });

    it('should log error on login failure', () => {
      mockSignIn.mockImplementation(() => {
        throw new Error();
      });
      render(<Header />);
      const loginButton = screen.getByText('Login');

      loginButton.click();

      expect(mockConsoleError).toBeCalled();
    });
  });

  describe('with logged in', () => {
    beforeEach(() => {
      useAuthenticationContext.mockReturnValue({ isLoggedIn: true });
    });

    it('should render correctly', () => {
      const tree = renderer.create(<Header />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should logout when logout button clicked', () => {
      render(<Header />);
      const logoutButton = screen.getByText('Logout');

      logoutButton.click();

      expect(mockSignOut).toBeCalled();
    });

    it('should log error on logout failure', () => {
      mockSignOut.mockImplementation(() => {
        throw new Error();
      });
      render(<Header />);
      const logoutButton = screen.getByText('Logout');

      logoutButton.click();

      expect(mockConsoleError).toBeCalled();
    });
  });
});
