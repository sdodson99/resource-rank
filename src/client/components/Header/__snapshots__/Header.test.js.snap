// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> with logged in should render correctly 1`] = `
<header
  className="Header"
  data-testid="Header"
>
  <div
    className="content-container"
  >
    <div
      className="sm:flex items-center text-center sm:text-left"
    >
      <div
        className="flex-grow"
      >
        <a
          href="/"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <div
            style={
              Object {
                "boxSizing": "border-box",
                "display": "inline-block",
                "height": 50,
                "overflow": "hidden",
                "position": "relative",
                "width": 50,
              }
            }
          >
            <img
              alt="Resource Rank Logo"
              decoding="async"
              src="/_next/image?url=%2Fimg%2Flogo.svg&w=128&q=75"
              srcSet="/_next/image?url=%2Fimg%2Flogo.svg&w=64&q=75 1x, /_next/image?url=%2Fimg%2Flogo.svg&w=128&q=75 2x"
              style={
                Object {
                  "border": "none",
                  "bottom": 0,
                  "boxSizing": "border-box",
                  "display": "block",
                  "height": 0,
                  "left": 0,
                  "margin": "auto",
                  "maxHeight": "100%",
                  "maxWidth": "100%",
                  "minHeight": "100%",
                  "minWidth": "100%",
                  "objectFit": undefined,
                  "objectPosition": undefined,
                  "padding": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "width": 0,
                }
              }
            />
          </div>
        </a>
      </div>
      <div
        className="NavItem NavLink Active"
      >
        <a
          href="/"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          Home
        </a>
      </div>
      <div
        className="NavItem NavLink"
      >
        <a
          href="/topics"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          Topics
        </a>
      </div>
      <div
        className="NavItem"
      >
        <button
          className="LoadingButton"
          data-testid="LoadingButton"
          onClick={[Function]}
        >
          Logout
        </button>
      </div>
    </div>
  </div>
</header>
`;

exports[`<Header /> with not logged in should render correctly 1`] = `
<header
  className="Header"
  data-testid="Header"
>
  <div
    className="content-container"
  >
    <div
      className="sm:flex items-center text-center sm:text-left"
    >
      <div
        className="flex-grow"
      >
        <a
          href="/"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <div
            style={
              Object {
                "boxSizing": "border-box",
                "display": "inline-block",
                "height": 50,
                "overflow": "hidden",
                "position": "relative",
                "width": 50,
              }
            }
          >
            <img
              alt="Resource Rank Logo"
              decoding="async"
              src="/_next/image?url=%2Fimg%2Flogo.svg&w=128&q=75"
              srcSet="/_next/image?url=%2Fimg%2Flogo.svg&w=64&q=75 1x, /_next/image?url=%2Fimg%2Flogo.svg&w=128&q=75 2x"
              style={
                Object {
                  "border": "none",
                  "bottom": 0,
                  "boxSizing": "border-box",
                  "display": "block",
                  "height": 0,
                  "left": 0,
                  "margin": "auto",
                  "maxHeight": "100%",
                  "maxWidth": "100%",
                  "minHeight": "100%",
                  "minWidth": "100%",
                  "objectFit": undefined,
                  "objectPosition": undefined,
                  "padding": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "width": 0,
                }
              }
            />
          </div>
        </a>
      </div>
      <div
        className="NavItem NavLink Active"
      >
        <a
          href="/"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          Home
        </a>
      </div>
      <div
        className="NavItem NavLink"
      >
        <a
          href="/topics"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          Topics
        </a>
      </div>
      <div
        className="NavItem"
      >
        <button
          className="LoadingButton"
          data-testid="LoadingButton"
          onClick={[Function]}
        >
          Login
        </button>
      </div>
    </div>
  </div>
</header>
`;
